daemon/jitter_buffer.c

```diff
@@ -0,0 +1,335 @@
+
+	return clock_rate;
+}
+
+static struct codec_packet* get_codec_packet(struct media_packet *mp, str *s) {
+	struct codec_packet *p = g_slice_alloc0(sizeof(*p));
+	p->s = *s;
+	p->packet = g_slice_alloc0(sizeof(*p->packet));
+	p->packet->sfd = mp->sfd;
+	p->packet->fsin = mp->fsin;
+	p->packet->tv = mp->tv;
+	p->free_func = free;
+	p->packet->buffered =1;
+
+	return p;
+}
+
+static void check_buffered_packets(struct jitter_buffer *jb, unsigned int len) {
+	if(len >= (2* jb_config->max_jb_len)) {
+		ilog(LOG_DEBUG, "Jitter reset due to buffer overflow");
+		mutex_lock(&jb->lock);
+		reset_jitter_buffer(jb);
+		mutex_unlock(&jb->lock);
+	}
+}
+
+static int queue_packet(struct media_packet *mp, struct codec_packet *p) {
+	struct jitter_buffer *jb = &mp->stream->rtp_sink->jb;
+	unsigned long ts = ntohl(mp->rtp->timestamp);
+	int payload_type =  (mp->rtp->m_pt & 0x7f);
+	int clockrate = get_clock_rate(mp, payload_type);
+	int ret = 0;
+
+	if(!clockrate || !jb->first_send.tv_sec) {
+		reset_jitter_buffer(jb);
+		mutex_unlock(&jb->lock);
+		p->packet->buffered = 0;
+		play_buffered(mp->stream->rtp_sink, p);
+		return 1;
+	}
+	long ts_diff = (uint32_t) ts - (uint32_t) jb->first_send_ts;
+	int seq_diff = ntohs(mp->rtp->seq_num) - jb->first_seq;
+	if(!jb->rtptime_delta) {
+		jb->rtptime_delta = ts_diff/seq_diff;
+	}
+	p->to_send = jb->first_send;
+	long long ts_diff_us =
+		(long long) (ts_diff + (jb->rtptime_delta * jb->buffer_len))* 1000000 / clockrate;
+
+	ts_diff_us += (jb->clock_drift_val * seq_diff); 
+
+	if(jb->buf_decremented) {
+		ts_diff_us += 5000; //add 5ms delta when 2 packets are scheduled around same time
+		jb->buf_decremented = 0;
+	}
+	timeval_add_usec(&p->to_send, ts_diff_us);
+
+	ts_diff_us = timeval_diff(&p->to_send, &rtpe_now);
+
+	if (ts_diff_us > 3000000) { // more than three second, can't be right
+		jb->first_send.tv_sec = 0;
+		jb->rtptime_delta = 0;
+	}
+
+	mutex_unlock(&jb->lock);
+	mutex_lock(&mp->stream->rtp_sink->out_lock);
+	g_queue_push_tail(&mp->packets_out, p);
+	ret =  media_socket_dequeue(mp, mp->stream->rtp_sink);
+	mutex_unlock(&mp->stream->rtp_sink->out_lock);
+
+	return ret;
+}
+
+static void handle_clock_drift(struct media_packet *mp) {
+	ilog(LOG_DEBUG, "handle_clock_drift");
+	unsigned long ts = ntohl(mp->rtp->timestamp);
+	struct jitter_buffer *jb = &mp->stream->rtp_sink->jb;
+	int payload_type =  (mp->rtp->m_pt & 0x7f);
+	int clockrate = get_clock_rate(mp, payload_type);
+	if(!clockrate) {
+		return;
+	}
+	long ts_diff = (uint32_t) ts - (uint32_t) jb->first_send_ts;
+	int seq_diff = ntohs(mp->rtp->seq_num) - jb->first_seq;
+	long long ts_diff_us =
+		(long long) (ts_diff)* 1000000 / clockrate;
+	struct timeval to_send = jb->first_send;
+	timeval_add_usec(&to_send, ts_diff_us);
+	long long time_diff = timeval_diff(&rtpe_now, &to_send);
+
+	jb->clock_drift_val = time_diff/seq_diff;
+	jb->clock_drift_enable = 0;
+	jb->cont_buff_err = 0;
+}
+
```

Comment:
Can you explain the algorithm used here? AFAICT this is a single-shot calculation that is done when an overflow or underflow occurs?